Simple wide integer implementation

mihails-strasuns-sociomantic opened the pull request:
Bare minimum I could test that does anything useful, available for early feedback.

mihails-strasuns-sociomantic committed:
Core WideUInt implementation

mihails-strasuns-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+    public double toDouble ( )
+    {
+        // TODO: for widenumbers with value > ulong.max need to calculate
+        // mantiss and exponent manually by counting bitshift
```
@don-clugston-sociomantic do we have anything in `ocean.math` to use for this?

joseph-wakeling-sociomantic replied:
Note, we need this `TODO` to be addressed before this satisfies our planned use-case (although, see above regarding the idea of allowing different word sizes).

mihails-strasuns-sociomantic replied:
Acknowledged, I only skipped it for now to see if I can learn from Don wisdom before getting into manual bit tinkering :)

joseph-wakeling-sociomantic replied:
I thoroughly endorse this approach :-)

david-eckardt-sociomantic replied:
There is [`ldexp(x, y) := x * 2^y`](http://pubs.opengroup.org/onlinepubs/9699919799/functions/ldexp.html) which you can probably use like this:
```d
double result = 0;
foreach (i, word; this.payload)
    result += ldexp(word, i * 32);
```

david-eckardt-sociomantic replied:
I also suggest using `real`.

mihails-strasuns-sociomantic replied:
My understanding is that `real` is to be avoided as much as possible unless one has a case with no other choice. We should ask FP master @don-clugston-sociomantic

don-clugston-sociomantic replied:
If you know that the payload length is always small, you can just use the brute force approach.

```d
double result = 0;
foreach_reverse (word, this.payload )
{
   result *= 2^^32; // this is an exactly representable number, it's an error-free calculation
   result += word;
}
```

If that's too slow you can speed it up by scanning for the first non-zero payload word. There are only 63 bits of mantissa in `real` so additions after the first three have no effect and can be skipped. (We need 3 words, not 2, because we might only get 1 bit from the first payload word).
If `p` is the index of the first non-zero word, then

```d
const double MAXWORD = 2.0^^32; // so we get floating-point multiplies. Could also use real instead`
return (((this.payload[p]*MAXWORD +(this.payload[p-1]))*MAXWORD)+this.payload[p-2])* ldexp(1.0, (p-2)*32);
```

Of course you need to worry about array bounds errors if p is less than 2. But you can set `p = min(p, 2)` because the formula is still valid even if the high payload values are zero.

david-eckardt-sociomantic replied:
It should be mentioned that `this.payload[p]*MAXWORD` is equivalent to `ldexp(this.payload[p], 32)`, but multiplication is faster than `ldexp`.

don-clugston-sociomantic replied:
Clarification, after discussing with David:
Note that `ldexp` is a slow operation, that is why I've only used it once in the code above.

gavin-norman-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+
+    ***************************************************************************/
+
+    this ( ulong value )
```
Huh? I've never seen this in D1!

mihails-strasuns-sociomantic replied:
You don't use structs a lot :) They do have constructors in D1, only destructors are missing.

joseph-wakeling-sociomantic replied:
Damn, this is news to me too.  I think I'd have used struct constructors more, if I'd known ... :-P

mihails-strasuns-sociomantic replied:
Hm, actually travis build rejects it. I may have some hacked dmd1 build locally :oops:

david-eckardt-sociomantic replied:
http://digitalmars.com/d/1.0/struct.html claims D1 structs don't have constructors (I wish they would).

gavin-norman-sociomantic replied:
I do use structs a lot! Haha. D2 only feature ;)

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+    this ( ulong value )
+    {
+        this.payload[0] = cast(uint) value;
+        this.payload[1] = value >> 32;
```
This is implicitly taking advantage of the fact that `payload[i]` will be default-initialized to 0 for i > 1, right?

mihails-strasuns-sociomantic replied:
Yes

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+    /***************************************************************************
+
+        Returns:
+            amount of meaninfful decimal digits in currently stored value
```
`meaningful`

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+    public void opAssign ( ulong rhs )
+    {
+        this.payload[] = 0;
+        this.payload[0] = cast(uint) (rhs & uint.max);
```
This is different to the constructor's assignment for `this.payload[0]`.  Is one of them wrong?

gavin-norman-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+        "For 'N <= 2' using 'uint' or 'ulong' directly is suggested"
+    );
+
+    private uint[N] payload;
```
Please document the ordering. It appears to be in ascending order of significance.

joseph-wakeling-sociomantic replied:
I guess it is more convenient to do it this way, but it does occur to me that we've got an interesting mix of big- and little-endian in terms of how this data is structured :-)

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+
+    this ( ulong value )
+    {
+        this.payload[0] = cast(uint) value;
```
This is different to how `opAssign ( ulong )` handles `this.payload[0]`.  Is one of them wrong?

mihails-strasuns-sociomantic replied:
No, just different ways to write the same operation. I will sync them.

joseph-wakeling-sociomantic replied:
Might be worth implementing some little internal set-the-payload function to DRY up things here?

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+
+    ***************************************************************************/
+
+    public uint divideBy ( uint rhs )
```
Having the remainder is nice, but is there a reason to not implement `opDivAssign`?

joseph-wakeling-sociomantic replied:
In addition, is there any reason not to implement division by another `WideInt`?

mihails-strasuns-sociomantic replied:
No, this is simply a most basic functional primitive to build other API on top

As for `WideInt rhs` overloads - do we actually need them now? I assumed it will be used as sort of aggregator value which only works with regular `ulong`/`uint` values individually but may internally grow over their size limit.

My intention is to implement minimal amount of functionality that fits current purpose and leave rest to PRs of interested parties ;)

joseph-wakeling-sociomantic replied:
As long as it's sufficient for the use-case I described to you (whose fine details I imagine by now you know better than I do:-), all good.

gavin-norman-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+    public void opAddAssign ( uint rhs )
+    {
+        if (add(this.payload[0], rhs))
+            enforce(.wideint_exception, checkAndInc(1));
```
`this.checkAndInc`

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+
+    ***************************************************************************/
+
+    public void multiplyBy ( uint rhs )
```
Is there a reason to not implement this as `opMulAssign`?  And is there a reason to not have an overload taking a `ulong`, or another `WideInt`?

(I appreciate that last one opens a can of worms in terms of, what _size_ of `WideInt`, but we could presumably consider it in terms of `WideInt` of same size to start with, and follow up with support for multiple different sizes.)

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+    {
+        WideInt num = 1_000_000_000;
+        num.multiplyBy(1_000_000_000);
+        test!("==")(num.toString(), "1000000000000000000");
```
Test the overflow exception, maybe?

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+import ocean.core.Enforce;
+
+/// ditto
+public struct WideInt ( size_t N )
```
With no signedness in place, isn't this strictly speaking a `WideUInt`?

mihails-strasuns-sociomantic replied:
Yes, I omitted it for aesthetical purpose. Do you think it is confusing?

joseph-wakeling-sociomantic replied:
I think it'd be better to be precise (especially as we might want to later add a genuine `WideInt` with signedness included; cf. how `BigUInt` and `BigInt` relate, although in that case the former is really treated more as an implementation detail of the latter.)

don-clugston-sociomantic replied:
`BigUint` is a misnomer. Fortunately it is not exposed.  I'd rather not perpetuate my earlier mistake.

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+/*******************************************************************************
+
+    Implementation of integer-like struct with arbitrary but fixed size, defined
```
`unsigned integer-like`, going by the current implementation ... ?

don-clugston-sociomantic replied:
To me, it looks like implementation of positive integers, not unsigned. It doesn't use wraparound semantics.

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+            overflow >>= 32;
+        }
+
+        enforce(.wideint_exception, overflow == 0);
```
Given that this is a wide unsigned integer, do we want this exception-on-overflow behaviour, or do we want unsigned integer-style wraparound (for consistency)?

The current behaviour looks simpler to implement, so I'm inclined to go with this personally.

mihails-strasuns-sociomantic replied:
I am quite confident we want exception in this case. Native integers have wraparound because it is how hardware works and it is fast. In almost all `WideInt` use cases you assume that it is "big enough" for all your values, essentually using it as `BigInt` with fixed size. If such assumption is violated, you want to know about it.

david-eckardt-sociomantic replied:
Well, hardware reports wrap-around as well, though the carry and overflow flags, but D swallows this exception.

mihails-strasuns-sociomantic replied:
No, it is not about carry (you may notice below that I have addition function that does report carry flag), it is about multiplication. When you multiply `ulong` by `ulong` your result can be up to 128bit large.

david-eckardt-sociomantic replied:
But that's overflow then, which is still reported by hardware but not D.

mihails-strasuns-sociomantic replied:
AFAIK hardware reports that it has happened, not overflown amount. That is sufficient for addition (where amount always == 1) but not multiplication.

david-eckardt-sociomantic replied:
That's right.

mihails-strasuns-sociomantic replied:
So you do agree that using `uint` for a single word is necessary? :)

david-eckardt-sociomantic replied:
Oh yes.

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+        "For 'N <= 2' using 'uint' or 'ulong' directly is suggested"
+    );
+
+    private uint[N] payload;
```
Do we want to insist on a `uint` wordsize, or also make that a template parameter of `WideInt` (perhaps defaulting to `uint` but not constrained to it)?

Motivation: assuming we want a 128-bit unsigned integer, wouldn't a `WideInt!(2, ulong)` be a more performant way of doing this (at least on a 64-bit system)?

mihails-strasuns-sociomantic replied:
AFAICS `uint` is the only option that makes sense. It allows to do multiplication of words by promoting them to `ulong` and handling overflow using regular arithmetic, with `ulong` you will effectively have to emulate CPU arithemtic unit yourself, hardly an efficient approach. Thus for 64bit `uint` is optimal unless you have hardware-optimized `cent` arithmetic available.

joseph-wakeling-sociomantic replied:
OK, makes sense.  Thanks for clarifying.  Might be worth noting in the documentation somewhere?

david-eckardt-sociomantic replied:
So the reason to use 32-bit words on 64-bit hardware is to handle carry information?

mihails-strasuns-sociomantic replied:
https://github.com/sociomantic-tsunami/ocean/pull/139#discussion_r119687711

don-clugston-sociomantic replied:
FYI: Some compilers have intrinsics to support these operations. Here are the gcc ones:
https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html
D should really add something like this as well.

gavin-norman-sociomantic reviewed (commented):
Couple of small noticings after a quick review.

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+        auto n = this.digits();
+        static mstring buffer;
+        buffer.length = n;
+        enableStomping(buffer);
```
This looks like a workaround for `sink` not accepting data one `char` at a time?  Or is that considered sufficiently inefficient that you want to capture the whole number in a single string _before_ passing it on to the sink?

What concerns me here is that this implementation eagerly allocates an entire array for formatting the number, rather than allowing `sink` to decide the allocation strategy.

mihails-strasuns-sociomantic replied:
No, it is because dividing by 10 will give your individual digits in reverse order compared to one you want to print. Maybe there are some algorithms to get digits starting with most significant one but I haven't bothered to search right now.

joseph-wakeling-sociomantic replied:
Ouch.  OK, fair enough; let's roll with this for now and we can follow up later if we identify a better way to do it.

don-clugston-sociomantic replied:
AFAIK in general there aren't any good algorithms for getting the digits in left-to-right order. The most efficient algorithms use even powers of 10, eg first divide by 100000000, then by 10000, then by 100. You can use that to give you the digits in a funny order rather than pure right-to-left, but it doesn't reduce the size of your buffer very much.

The best you can do is divide by 10^19 first (since that tightly fits into a `ulong`). Then you can produce the output in chunks of 19 digits, instead of 1 character at a time. If your base type is `uint` then you can only use 10^9.
That reduces the required buffer space by a factor of more than 2. But since it does not eliminate it, it's rather marginal benefit.

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+
+    public void opAddAssign ( uint rhs )
+    {
+        if (add(this.payload[0], rhs))
```
I'm not sure I quite follow how this `add` function is actually working in practice.  Can you explain the magic? :-)

mihails-strasuns-sociomantic replied:
It adds `rhs` to `this.payload[0]` (with regular unsigned int wraparound semantics) and returns `true` if wraparound has happened - which means higher order digit needs to be incremented.

joseph-wakeling-sociomantic replied:
Oh, I see.  I think I just didn't look properly at what `checkAndInc` was doing before, so didn't follow either the consequences or the reason for the `enforce`.  Thanks again for clarifying.

joseph-wakeling-sociomantic reviewed (commented):
Looks pretty nice!  I have a few thoughts/remarks/questions ...

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+            return false;
+
+        if (this.payload[1] != (rhs >> 32))
+            return false;
```
This is where I miss the opportunity to return a `uint[2]` from some helper function.  We could encapsulate everything needed to assign from `ulong`, and also generate a single point of comparison for `this.payload[0 .. 2]` ... :-P

joseph-wakeling-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+
+    ***************************************************************************/
+
+    public uint divideBy ( uint rhs )
```
As long as it's sufficient for `BidYieldProfile`'s needs (which I imagine you've already looked at in more fine detail than me:-), all good.

david-eckardt-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+        {
+            auto remainder = copy.divideBy(10);
+            buffer[idx] = cast(char) ('0' + remainder);
+        }
```
Sorry, but I can't help myself:
```d
foreach_reverse (ref c; buffer)
{
    c = ...
}
```
is just easier to read than
```d
for (ptrdiff_t idx = n-1; idx >= 0; --idx)
{
    buffer[idx] = ...
}
```
and inherently safe from off-by-one and index overflow errors.

david-eckardt-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+    {
+        WideInt num = 1_222_333_444;
+        num.multiplyBy(1_000_000_000);
+        test!("is")(num.toDouble(), 1_222_333_444e+9);
```
You are sure you won't get in trouble with limited floating-point precision, right? :trollface:

mihails-strasuns-sociomantic replied:
Airthmetic here macthes internal floating point representation exactly (fits the mantiss and only changes exponent) so I would expect result to be binary identical (thus `is`).

david-eckardt-sociomantic replied:
Do you know that for sure? 1,222,333,444,000,000,000 is divisible by 2048 so this test will still pass if you lose 11 mantissa bits and may fail for 1,222,333,444,000,000,001 or even 1,222,333,444,100,000,000.
Ok, you *are* right, but maybe verify this by something like
```d
static assert(1_222_333_444_000_000_000 > (1 << double.mant_dig));
```

david-eckardt-sociomantic replied:
> Airthmetic here macthes internal floating point representation exactly (fits the mantiss and only changes exponent

Oh no. You specify a _decimal_ exponent here, but IEEE754 floating-point values use _binary_ exponents: The exact FP representation is `mantissa * 2^exponent` (where `mantissa` is a fixed-point fractional number between `1.000...` and `1.111...`).
There are two ways (I'm aware of) of constructing exact FP values:
- at compile time: the [hexadecimal floating-point number literal](http://dlang.org/spec/lex.html#HexFloat)
- at run time: [`ldexp(a, b)`](http://pubs.opengroup.org/onlinepubs/9699919799/functions/ldexp.html) where `a` is an integer less than `2^double.mant_dig`.

mihails-strasuns-sociomantic replied:
You are right, I have been thinking about it as if it has `0b1000_ ...`. Will need to rework this test case completely.

david-eckardt-sociomantic replied:
To clarify: The decimal `{x}.{y}e{z}` notation is for convenience only and unrelated to the actual (binary exponential) representation of the resulting value.

david-eckardt-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+
+    ***************************************************************************/
+
+    public size_t digits ( )
```
Better: `decimal_digits`

hatem-oraby-sociomantic replied:
would make sense to do it as: `size_t digits (base = 10)` to make it more generic?

david-eckardt-sociomantic replied:
That is tempting (Tango has such string formatting) but I don't think it's worth it. Except for decimal I've only ever seen and used bases of powers of 2, and they come down to `lzcnt`.

nemanja-boric-sociomantic replied:
This method could use an unittest.

nemanja-boric-sociomantic commented on src/ocean/math/WideInt.d:
```diff
@@ -0,0 +1,474 @@
+    {
+        WideInt copy = *this;
+        size_t count;
+        while (copy != 0)
```
This will fail for `WideInt num = 0`; it will report 0 digits, but it has 1 digit.

mihails-strasuns-sociomantic committed:
Deprecate BigInt

mihails-strasuns-sociomantic commented:
Updated with documentation fixes, renamed to `WideUInt` and lot of test improvements.
`toDouble` now uses implementation suggested by @don-clugston-sociomantic though it could use test case suggestions too ;)

mihails-strasuns-sociomantic commented:
== I think this is ready for review/merge as initial prototype at least

gavin-norman-sociomantic commented on src/ocean/math/WideUInt.d:
```diff
@@ -0,0 +1,591 @@
+
+    private uint[N] payload;
+
+    version (D_Version2)
```
D2-only behaviour?

mihails-strasuns-sociomantic replied:
Nope, this is required for compatibility between D1 and D2. WideUInt num = 42 will result in opAssign being called in D1 but in D2 it requires constructor.

gavin-norman-sociomantic replied:
Fair enough.

mihails-strasuns-sociomantic commented:
Ping

joseph-wakeling-sociomantic commented:
Amusing coincidence:
https://forum.dlang.org/thread/fqkwcjhlnhrdgtikwhue@forum.dlang.org

(There are a couple of different implementations that other people are doing linked in there)

nemanja-boric-sociomantic commented on src/ocean/math/WideUInt.d:
```diff
@@ -0,0 +1,591 @@
+            // code output should be checked to see if assembly block can be
+            // reliably replaced with plain D arithmetics
+
+            asm
```
I would guess this would be pretty close:

```
static uint add ( uint a, uint b, out bool overflow )
{
    ulong res = cast(ulong)(a) + cast(ulong)(b);
    overflow = res >> 0x20;
    return cast(uint)res;
}
```

david-eckardt-sociomantic replied:
I like using the CPU's built-in overflow detection.

nemanja-boric-sociomantic replied:
Most (all?) of the modern CPUs do have barrel shifter with which this operation costs nothing, and I would guess (though it's just my gut feeling, I have no numbers) it's not slower than using non-portable (not even across the compilers) CPU instructions.

mihails-strasuns-sociomantic replied:
I played a bit with d.godbolt.org and it looks like both gdc and ldc generate suboptimal code for this add: https://godbolt.org/g/07RZO2

However, there is alternative implementation which result in exactly same assembly:

```D

uint add2 ( uint a, uint b, out bool overflow )
{
    uint result = a + b;
    overflow = result < a;
    return result;
}
```
I will use it as a fallback.

mihails-strasuns-sociomantic commented:
Updated with D code fallback for assembly block

mihails-strasuns-sociomantic commented:
Ping

nemanja-boric-sociomantic commented on relnotes/wideuint.feature.md:
```diff
@@ -0,0 +1,8 @@
+* `ocean.math.WideUint`
+
+  New module that implements positive integer number of a fixed size. It is
```
`s/positive/non-negative`?

nemanja-boric-sociomantic commented on relnotes/bigint.deprecation.md:
```diff
@@ -0,0 +1,7 @@
+* `ocean.math.BigInt`
+
+  This module is deprecated in favor of `ocean.math.WideUint` because its
```
Given the open source position on this, this release note seems a little bit too brutal. For example, `WideUint` is not a replacement for a `BigInt`. I'm not sure we can do anything about it, though.

nemanja-boric-sociomantic commented on src/ocean/math/WideUInt.d:
```diff
@@ -0,0 +1,591 @@
+    binary sequence representing actual stored number.
+
+    Copyright:
+        Copyright (c) 2017 Sociomantic Labs GmbH.
```
lowercase `s` in `sociomantic` Sorry :-)

nemanja-boric-sociomantic commented on src/ocean/math/WideUInt.d:
```diff
@@ -0,0 +1,591 @@
+import ocean.math.IEEE : feqrel;
+
+/// ditto
+public struct WideUInt ( size_t N )
```
Missing documentation for `N`

nemanja-boric-sociomantic reviewed (commented):
Generally LGTM. @david-eckardt-sociomantic, @gavin-norman-sociomantic ?

gavin-norman-sociomantic commented:
You can probably update the commit message from "Start of WideUInt implementation" now ;)

gavin-norman-sociomantic commented on relnotes/wideuint.feature.md:
```diff
@@ -0,0 +1,8 @@
+* `ocean.math.WideUint`
```
`WideUInt` or `WideUint`?

gavin-norman-sociomantic commented on src/ocean/math/WideUInt.d:
```diff
@@ -0,0 +1,591 @@
+/*******************************************************************************
+
+    Implementation struct emulating positive number with value range exceeding
```
This is clearer:
"Struct emulating a large (bigger than ulong) non-negative integer of fixed range (defined via template argument)."

gavin-norman-sociomantic commented on src/ocean/math/WideUInt.d:
```diff
@@ -0,0 +1,591 @@
+    ***************************************************************************/
+
+    public uint divideBy ( uint rhs )
+    {
```
You could add an out contract that `remainder < rhs`.

gavin-norman-sociomantic commented on src/ocean/math/WideUInt.d:
```diff
@@ -0,0 +1,591 @@
+
+        Params:
+            rhs = value to compare to
+
```
Returns:

gavin-norman-sociomantic commented on src/ocean/math/WideUInt.d:
```diff
@@ -0,0 +1,591 @@
+        test!("!=")(a, b);
+    }
+
+    /**************************************************************************/
```
Why no doc?

gavin-norman-sociomantic commented on src/ocean/math/WideUInt.d:
```diff
@@ -0,0 +1,591 @@
+        assert(remainder <= uint.max);
+        return cast(uint) remainder;
+    }
+
```
`alias opDivAssign divideBy;`?

mihails-strasuns-sociomantic replied:
Omitted intentionally for now. I would prefer to add "meat to bones" (various opeator overloads) separately from core impl.

gavin-norman-sociomantic replied:
Fair enough.

gavin-norman-sociomantic commented on src/ocean/math/WideUInt.d:
```diff
@@ -0,0 +1,591 @@
+
+        enforce(.wideint_exception, overflow == 0);
+    }
+
```
`alias opMulAssign multiplyBy;`?

gavin-norman-sociomantic commented on src/ocean/math/WideUInt.d:
```diff
@@ -0,0 +1,591 @@
+
+        Returns:
+            double precision floating point representation of this number
+
```
What happens if the integer is larger than can be represented by a double? (The maximum double value is 10^308, which could be stored in a 1024-bit unsigned-integer, so a very small `WideUInt`.)

don-clugston-sociomantic replied:
It will be double.infinity. How about:
```
Returns:
double precision floating point value which is nearest to this number.
Uses the same semantics as conversion from ulong to double.
```

mihails-strasuns-sociomantic replied:
Good wording!

gavin-norman-sociomantic reviewed (commented):
Focusing on doc and API...

mihails-strasuns-sociomantic commented:
Updated.
